// konsserto/lib/src/Konsserto/Component/Form/FormFactory.js

/*
 * This file is part of the Konsserto package.
 *
 * (c) Jessym Reziga <jessym@konsserto.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
var Form, FormFactory, Tools;

Form = use('@Konsserto/Component/Form/Form');

Tools = use('@Konsserto/Component/Static/Tools');

FormFactory = (function() {
  function FormFactory(stackRequest, templating, entityManager) {
    this.stackRequest = stackRequest;
    this.templating = templating;
    this.entityManager = entityManager;
    this.formItem = [];
  }

  FormFactory.prototype.createBuilder = function(name, data, options) {
    var form;
    if (name == null) {
      name = 'form';
    }
    if (data == null) {
      data = {};
    }
    if (options == null) {
      options = {};
    }
    form = new Form(name, data, options, this.stackRequest.getCurrentRequest(), this.templating, this.entityManager);
    this.formItem[name] = form;
    return form;
  };

  return FormFactory;

})();

module.exports = FormFactory;
