// konsserto/lib/src/Konsserto/Component/Finder/Finder.js

/*
 * This file is part of the Konsserto package.
 *
 * (c) Marvin Frachet <marvin@konsserto.com>
 * (c) Jessym Reziga <jessym@konsserto.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
var DirectoryFinder, FileFinder, Finder, fs;

fs = use('fs');

DirectoryFinder = use('@Konsserto/Component/Finder/DirectoryFinder');

FileFinder = use('@Konsserto/Component/Finder/FileFinder');

Finder = (function() {
  function Finder() {

    /*
      		Instanciation of a new Finder Service
     */
    this.isRecursive = false;
    return this;
  }

  Finder.prototype.files = function() {

    /*
      		Make the choice to manage files with the finder
     */
    this.fileFinder = new FileFinder();
    return this;
  };

  Finder.prototype.recursive = function() {

    /*
      		Make the choice to make a recursive search
     */
    this.isRecursive = true;
    return this;
  };

  Finder.prototype.notRecursive = function() {

    /*
      		Make the choice to explicitly don't make a recursive search
     */
    this.isRecursive = false;
    return this;
  };

  Finder.prototype.directories = function() {

    /*
      		Make the choice to manage directories with the finder
     */
    this.directoryFinder = new DirectoryFinder();
    return this;
  };

  Finder.prototype["in"] = function(rootDir, recursive) {
    if (recursive == null) {
      recursive = this.isRecursive;
    }

    /*
      		Get the list of the choosen type in the current folder
     */
    if (fs.existsSync(rootDir)) {
      if (this.fileFinder) {
        if (!recursive) {
          this.fileFinder.getFiles(rootDir);
        } else {
          this.fileFinder.getRecursive(rootDir);
        }
      } else {
        if (this.directoryFinder) {
          this.directoryFinder.getDirs(rootDir, recursive);
        }
      }
    }
    return this;
  };

  Finder.prototype.display = function() {

    /*
      		Get the resulting list corresponding to the search
     */
    if (this.fileFinder) {
      return this.fileFinder.getResult();
    } else {
      if (this.directoryFinder.getResult().length === 0) {
        this.directoryFinder.getDirs('.', true);
      }
      return this.directoryFinder.getResult();
    }
  };

  Finder.prototype.name = function(name) {

    /*
      		Reduce the resulting list by the name parameter
     */
    if (this.fileFinder) {
      this.fileFinder.name(name);
    } else {
      this.directoryFinder.name(name);
    }
    return this;
  };

  Finder.prototype.notName = function(name) {

    /*
      		Reduce the resulting list if the name is different from the parameter
     */
    if (this.fileFinder) {
      this.fileFinder.notName(name);
    } else {
      this.directoryFinder.notName(name);
    }
    return this;
  };

  Finder.prototype.size = function(size) {

    /*
      		Get the file size
     */
    this.fileFinder.size(size);
    return this;
  };

  Finder.prototype.sortBy = function(orderby, orderway) {

    /*
      		Sort the resulting list
     */
    if (this.fileFinder) {
      this.fileFinder.sortBy(orderby, orderway);
    } else {
      this.directoryFinder.sortBy(orderby, orderway);
    }
    return this;
  };

  Finder.prototype.contains = function(content) {

    /*
      		Reduce the files list that don't contain the content parameter value"
     */
    if (this.fileFinder) {
      this.fileFinder.contains(content);
    }
    return this;
  };

  Finder.prototype.ext = function(extension) {

    /*
      		Reduce the files list that don't have the extension passed in paramter"
     */
    if (this.fileFinder) {
      this.fileFinder.ext(extension);
    }
    return this;
  };

  Finder.prototype.end = function() {
    return this.display();
  };

  return Finder;

})();

module.exports = Finder;
