// konsserto/lib/src/Konsserto/Component/Finder/File.js

/*
 * This file is part of the Konsserto package.
 *
 * (c) Marvin Frachet <marvin@konsserto.com>
 * (c) Jessym Reziga <jessym@konsserto.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
var File, fs;

fs = require("fs");

File = (function() {
  function File(name, relativePath) {
    this.name = name;
    this.relativePath = relativePath;
    if (this.relativePath.charAt(this.relativePath.length - 1) !== '/') {
      this.relativePath += '/';
    }
    this.relativePathName = this.relativePath + this.name;
    this.file = fs.statSync(this.relativePathName);
  }

  File.prototype.getSize = function() {
    return this.file.size;
  };

  File.prototype.getAddedTime = function() {
    return this.file.atime;
  };

  File.prototype.getUpTime = function() {
    return this.file.mtime;
  };

  File.prototype.getRelativePath = function() {
    return this.relativePath;
  };

  File.prototype.getRelativePathName = function() {
    return this.relativePathName;
  };

  File.prototype.getName = function() {
    return this.name;
  };

  File.prototype.getExtension = function() {
    var ext;
    ext = this.name.split('.').pop();
    if (ext === this.name) {
      return '';
    }
    return ext;
  };

  File.prototype.getContent = function() {
    return fs.readFileSync(this.relativePathName, {
      encoding: 'utf-8'
    });
  };

  File.prototype.getNameWithoutExtension = function() {
    return this.name.replace('.' + this.getExtension(), '');
  };

  return File;

})();

module.exports = File;
