// konsserto/lib/src/Konsserto/Bundle/TwigBundle/Extension/FilterExtension.js

/*
 * This file is part of the Konsserto package.
 *
 * (c) Jessym Reziga <jessym@konsserto.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
var TwigExtension, TwigFilterExtension, TwigFilterMethod,
  __bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; },
  __hasProp = {}.hasOwnProperty,
  __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

TwigExtension = use('@Konsserto/Vendor/Twig/Extension/TwigExtension');

TwigFilterMethod = use('@Konsserto/Vendor/Twig/Extension/TwigFilterMethod');

TwigFilterExtension = (function(_super) {
  __extends(TwigFilterExtension, _super);

  function TwigFilterExtension(translator) {
    this.translator = translator;
    this.echo = __bind(this.echo, this);
    this.trans = __bind(this.trans, this);
    this.getFilters = __bind(this.getFilters, this);
    TwigFilterExtension.__super__.constructor.call(this, constructor);
  }

  TwigFilterExtension.prototype.getFilters = function() {
    return [
      {
        name: 'trans',
        access: new TwigFilterMethod(this.trans)
      }, {
        name: 'echo',
        access: new TwigFilterMethod(this.echo)
      }
    ];
  };

  TwigFilterExtension.prototype.trans = function(id, args) {
    var domain, locale, parameters, _ref, _ref1, _ref2;
    parameters = (_ref = args[0]) != null ? _ref : null;
    domain = (_ref1 = args[1]) != null ? _ref1 : null;
    locale = (_ref2 = args[2]) != null ? _ref2 : null;
    return this.translator.trans(id, parameters, domain, locale);
  };

  TwigFilterExtension.prototype.echo = function(value, str) {
    return '<pre>Filtered Value : ' + value + ' | Echo : ' + str + '</pre>';
  };

  return TwigFilterExtension;

})(TwigExtension);

module.exports = TwigFilterExtension;
