// konsserto/lib/src/Konsserto/Component/Cache/Cache.js

/*
 * This file is part of the Konsserto package.
 *
 * (c) Marvin Frachet <marvin@konsserto.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
var CONFIG, Cache, node_cache;

node_cache = use("node-cache");

CONFIG = use('/app/config/config');

Cache = (function() {
  function Cache() {
    this.cache = new node_cache({
      checkperiod: CONFIG.cacheTime
    });
  }

  Cache.prototype.set = function(key, value) {
    return this.cache.set(key, value);
  };

  Cache.prototype.remove = function(key) {
    return this.cache.del(key);
  };

  Cache.prototype.removeMultiple = function(keys) {
    return this.cache.del(keys);
  };

  Cache.prototype.get = function(key) {
    return this.cache.get(key);
  };

  Cache.prototype.getMultiple = function(keys) {
    return this.cache.get(keys);
  };

  Cache.prototype.getStats = function() {
    return this.cache.getStats();
  };

  Cache.prototype.flushAll = function() {
    return this.cache.flushAll();
  };

  return Cache;

})();

module.exports = Cache;
