// konsserto/lib/src/Konsserto/Component/HttpKernel/RequestStack.js

/*
 * This file is part of the Konsserto package.
 *
 * (c) Jessym Reziga <jessym@konsserto.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
var RequestStack;

RequestStack = (function() {
  function RequestStack(translator) {
    this.translator = translator;
    this.requests = [];
  }

  RequestStack.prototype.push = function(request) {
    return this.requests.push(request);
  };

  RequestStack.prototype.pop = function() {
    if (!this.requests) {
      return null;
    }
    return this.requests.pop();
  };

  RequestStack.prototype.getCurrentRequest = function() {
    var len;
    len = this.requests.length - 1;
    if (this.requests[len]) {
      return this.requests[len];
    }
    return null;
  };

  RequestStack.prototype.getMasterRequest = function() {
    if (!this.requests) {
      return null;
    }
    return this.requests[0];
  };

  RequestStack.prototype.getParentRequest = function() {
    var pos;
    pos = this.requests.len - 2;
    if (this.requests[pos]) {
      return null;
    }
    return this.requests[pos];
  };

  return RequestStack;

})();

module.exports = RequestStack;
