// konsserto/lib/src/Konsserto/Component/Router/RouteInstance.js

/*
 * This file is part of the Konsserto package.
 *
 * (c) Jessym Reziga <jessym@konsserto.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
var RouteInstance, Tools;

Tools = use('@Konsserto/Component/Static/Tools');

RouteInstance = (function() {
  function RouteInstance(routeDefinition, controllerDefinition) {
    var route;
    this.routeDefinition = routeDefinition;
    this.controllerDefinition = controllerDefinition;
    this.controller = new controllerDefinition;
    route = this.getDefinition();
    console.log('[Route] ' + this.getControllerName() + ':' + this.getMethod() + ' | pattern: ' + route.getRawPattern());
  }

  RouteInstance.prototype.setController = function(controller) {
    this.controller = controller;
    return this;
  };

  RouteInstance.prototype.setRouteDefinition = function(routeDefinition) {
    this.routeDefinition = routeDefinition;
    return this;
  };

  RouteInstance.prototype.setRequest = function(request) {
    this.request = request;
    this.getController().setRequest(this.request);
    return this;
  };

  RouteInstance.prototype.setResponse = function(response) {
    this.response = response;
    this.getController().setResponse(this.response);
    return this;
  };

  RouteInstance.prototype.setControllerDefinition = function(controllerDefinition) {
    this.controllerDefinition = controllerDefinition;
    return this;
  };

  RouteInstance.prototype.getDefinition = function() {
    return this.routeDefinition;
  };

  RouteInstance.prototype.getController = function() {
    return this.controller;
  };

  RouteInstance.prototype.getControllerName = function() {
    return this.getController().constructor.name;
  };

  RouteInstance.prototype.getRequest = function() {
    return this.request;
  };

  RouteInstance.prototype.getResponse = function() {
    return this.response;
  };

  RouteInstance.prototype.setNext = function(next) {
    this.next = next;
    this.getController().setNext(this.next);
    return this;
  };

  RouteInstance.prototype.getNext = function() {
    return this.next;
  };

  RouteInstance.prototype.getControllerDefinition = function() {
    return this.controllerDefinition;
  };

  RouteInstance.prototype.getMethod = function() {
    return this.getDefinition().getMethodName() + 'Action';
  };

  RouteInstance.prototype.methodExists = function() {
    var method;
    method = this.getMethod();
    if ((method == null) || (this.getController()[method] == null)) {
      return false;
    }
    return true;
  };

  RouteInstance.prototype.getMethodArguments = function() {
    var arg, args, chainArgs, definition, params, _i, _len;
    chainArgs = [];
    definition = this.getDefinition();
    args = Tools.getFunctionParameters(this.getControllerDefinition().prototype[this.getMethod()]);
    params = this.getRequest().params;
    for (_i = 0, _len = args.length; _i < _len; _i++) {
      arg = args[_i];
      if (params[arg] != null) {
        chainArgs.push(params[arg]);
      } else if (definition.hasDefaultArg(arg)) {
        chainArgs.push(definition.getDefaultValueForArg(arg));
      } else {
        chainArgs.push(null);
      }
    }
    return chainArgs;
  };

  RouteInstance.prototype.getMethodInstance = function() {
    if (this.getController()[this.getMethod()] != null) {
      return this.getController()[this.getMethod()];
    }
    throw new Error('No method ' + this.getMethod() + ' in controller ' + this.getDefinition().getControllerName());
  };

  RouteInstance.prototype.getResponse = function() {
    return Tools.call(this.getMethodInstance(), this.getMethodArguments());
  };

  return RouteInstance;

})();

module.exports = RouteInstance;
