// konsserto/lib/src/Konsserto/Component/Middleware/MiddlewareInstance.js

/*
 * This file is part of the Konsserto package.
 *
 * (c) Jessym Reziga <jessym@konsserto.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
var MiddlewareInstance, Tools;

Tools = use('@Konsserto/Component/Static/Tools');

MiddlewareInstance = (function() {
  function MiddlewareInstance(middlewareDefinition, controllerDefinition) {
    this.middlewareDefinition = middlewareDefinition;
    this.controllerDefinition = controllerDefinition;
    this.controller = new controllerDefinition;
    console.log('[Middleware] middleware: ' + this.getControllerName() + ':' + this.getMethod());
  }

  MiddlewareInstance.prototype.setController = function(controller) {
    this.controller = controller;
    return this;
  };

  MiddlewareInstance.prototype.setMiddlewareDefinition = function(middlewareDefinition) {
    this.middlewareDefinition = middlewareDefinition;
    return this;
  };

  MiddlewareInstance.prototype.setControllerDefinition = function(controllerDefinition) {
    this.controllerDefinition = controllerDefinition;
    return this;
  };

  MiddlewareInstance.prototype.getDefinition = function() {
    return this.middlewareDefinition;
  };

  MiddlewareInstance.prototype.getController = function() {
    return this.controller;
  };

  MiddlewareInstance.prototype.getControllerName = function() {
    return this.getController().constructor.name;
  };

  MiddlewareInstance.prototype.getControllerDefinition = function() {
    return this.controllerDefinition;
  };

  MiddlewareInstance.prototype.getMethod = function() {
    return this.getDefinition().getMethodName() + 'Action';
  };

  MiddlewareInstance.prototype.getRequest = function() {
    return this.request;
  };

  MiddlewareInstance.prototype.getResponse = function() {
    return this.response;
  };

  MiddlewareInstance.prototype.setRequest = function(request) {
    this.request = request;
    this.getController().setRequest(this.request);
    return this;
  };

  MiddlewareInstance.prototype.setResponse = function(response) {
    this.response = response;
    this.getController().setResponse(this.response);
    return this;
  };

  MiddlewareInstance.prototype.setNext = function(next) {
    this.next = next;
    this.getController().setNext(this.next);
    return this;
  };

  MiddlewareInstance.prototype.getNext = function() {
    return this.next;
  };

  MiddlewareInstance.prototype.methodExists = function() {
    var method;
    method = this.getMethod();
    if ((method == null) || (this.getController()[method] == null)) {
      return false;
    }
    return true;
  };

  MiddlewareInstance.prototype.getMethodInstance = function() {
    if (this.getController()[this.getMethod()] != null) {
      return this.getController()[this.getMethod()];
    }
    throw new Error('No method ' + this.getMethod() + ' in middleware ' + this.getDefinition().getMiddlewareName());
  };

  MiddlewareInstance.prototype.getResponse = function() {
    return Tools.call(this.getMethodInstance(), {});
  };

  return MiddlewareInstance;

})();

module.exports = MiddlewareInstance;
