// konsserto/lib/src/Konsserto/Component/Middleware/Middleware.js

/*
 * This file is part of the Konsserto package.
 *
 * (c) Marvin Frachet <marvin@konsserto.com>
 * (c) Jessym Reziga <jessym@konsserto.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
var Middleware;

Middleware = (function() {
  function Middleware() {}

  Middleware.prototype.setContainer = function(container) {
    this.container = container;
    return this;
  };

  Middleware.prototype.setApplication = function(app) {
    this.app = app;
    return this;
  };

  Middleware.prototype.getRequest = function() {
    return this.container.get('Request_Stack').getCurrentRequest();
  };

  Middleware.prototype.getResponse = function() {
    return this.response;
  };

  Middleware.prototype.setRequest = function(request) {
    this.request = request;
    return this;
  };

  Middleware.prototype.setResponse = function(response) {
    this.response = response;
    return this;
  };

  Middleware.prototype.setNext = function(next) {
    this.next = next;
    return this;
  };

  Middleware.prototype.get = function(name) {
    return this.container.get(name);
  };

  Middleware.prototype.getContainer = function() {
    return this.container;
  };

  Middleware.prototype.getApplication = function() {
    return this.application;
    return {
      next: function() {
        return this.next();
      }
    };
  };

  return Middleware;

})();

module.exports = Middleware;
