// konsserto/lib/src/Konsserto/Component/Middleware/MiddlewareDefinition.js

/*
 * This file is part of the Konsserto package.
 *
 * (c) Jessym Reziga <jessym@konsserto.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
var MiddlewareDefinition;

MiddlewareDefinition = (function() {
  function MiddlewareDefinition(registeredMiddleware) {
    var middleware;
    this.registeredMiddleware = registeredMiddleware;
    middleware = this.registeredMiddleware.middleware.split(':');
    this.name = this.registeredMiddleware.name;
    this.bundleName = middleware[0];
    this.middlewareName = middleware[1];
    this.methodName = middleware[2];
  }

  MiddlewareDefinition.prototype.setBundleName = function(bundleName) {
    this.bundleName = bundleName;
    return this;
  };

  MiddlewareDefinition.prototype.setMiddlewareName = function(middlewareName) {
    this.middlewareName = middlewareName;
    return this;
  };

  MiddlewareDefinition.prototype.setMethodName = function(methodName) {
    this.methodName = methodName;
    return this;
  };

  MiddlewareDefinition.prototype.setName = function(name) {
    this.name = name;
    return this;
  };

  MiddlewareDefinition.prototype.getBundleName = function() {
    return this.bundleName;
  };

  MiddlewareDefinition.prototype.getMiddlewareName = function() {
    return this.middlewareName;
  };

  MiddlewareDefinition.prototype.getMethodName = function() {
    return this.methodName;
  };

  MiddlewareDefinition.prototype.getName = function() {
    return this.name;
  };

  return MiddlewareDefinition;

})();

module.exports = MiddlewareDefinition;
