// konsserto/lib/src/Konsserto/Component/ORM/EntityInstance.js

/*
 * This file is part of the Konsserto package.
 *
 * (c) Jessym Reziga <jessym@konsserto.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
var EntityInstance, EntityRepository;

EntityRepository = use('@Konsserto/Component/ORM/EntityRepository');

EntityInstance = (function() {
  function EntityInstance(entityDefinition, database, em) {
    var e, model, repositoryClass;
    this.entityDefinition = entityDefinition;
    this.database = database;
    this.em = em != null ? em : null;
    model = this.entityDefinition.getModel();
    this.modelInstance = this.database.define(model.getTableName(), model.getFields(), model.getOptions());
    repositoryClass = EntityRepository;
    if (model.hasRepository()) {
      try {
        repositoryClass = use(model.getRepositoryPath());
      } catch (_error) {
        e = _error;
        throw new Error('The repository path \'' + model.getRepositoryPath() + '\' doesn\'t exists.\n' + e.message);
      }
    }
    this.repository = new repositoryClass(this);
  }

  EntityInstance.prototype.getDatabase = function() {
    return this.database;
  };

  EntityInstance.prototype.getEntityManager = function() {
    return this.em;
  };

  EntityInstance.prototype.getDefinition = function() {
    return this.entityDefinition;
  };

  EntityInstance.prototype.setDefinition = function(entityDefinition) {
    this.entityDefinition = entityDefinition;
    return this;
  };

  EntityInstance.prototype.getModelInstance = function() {
    return this.modelInstance;
  };

  EntityInstance.prototype.setModelInstance = function(modelInstance) {
    this.modelInstance = modelInstance;
    return this;
  };

  EntityInstance.prototype.getRepository = function() {
    return this.repository;
  };

  EntityInstance.prototype.setRepository = function(repository) {
    this.repository = repository;
    return this;
  };

  return EntityInstance;

})();

module.exports = EntityInstance;
