// konsserto/lib/src/Konsserto/Component/ORM/EntityDefinition.js

/*
 * This file is part of the Konsserto package.
 *
 * (c) Jessym Reziga <jessym@konsserto.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
var EntityDefinition, Model;

Model = use('@Konsserto/Component/ORM/Model');

EntityDefinition = (function() {
  function EntityDefinition(rawFile, bundle, dirties) {
    this.rawFile = rawFile;
    this.bundle = bundle;
    this.name = this.rawFile.getNameWithoutExtension();
    this.logicalPath = this.bundle.getName() + ':' + this.name;
    this._class = this.bundle.fetchEntity(this.name);
    this.model = new Model(this.name, this._class, this.logicalPath, dirties);
  }

  EntityDefinition.prototype.setName = function(name) {
    this.name = name;
    return this;
  };

  EntityDefinition.prototype.setClass = function(_class) {
    this._class = _class;
    return this;
  };

  EntityDefinition.prototype.setRawFile = function(rawFile) {
    this.rawFile = rawFile;
    return this;
  };

  EntityDefinition.prototype.setBundle = function(bundle) {
    this.bundle = bundle;
    return this;
  };

  EntityDefinition.prototype.setLogicalPath = function(logicalPath) {
    this.logicalPath = logicalPath;
    return this;
  };

  EntityDefinition.prototype.setModel = function(model) {
    this.model = model;
    return this;
  };

  EntityDefinition.prototype.getName = function() {
    return this.name;
  };

  EntityDefinition.prototype.getClass = function() {
    return this._class;
  };

  EntityDefinition.prototype.getRawFile = function() {
    return this.rawFile;
  };

  EntityDefinition.prototype.getBundle = function() {
    return this.bundle;
  };

  EntityDefinition.prototype.getLogicalPath = function() {
    return this.logicalPath;
  };

  EntityDefinition.prototype.getPath = function() {
    return this._class._fromUsed;
  };

  EntityDefinition.prototype.getModel = function() {
    return this.model;
  };

  return EntityDefinition;

})();

module.exports = EntityDefinition;
