// konsserto/lib/src/Konsserto/Component/EntityManager/EntityManager.js
var EntityManager, FactoryAdapter, cc;

cc = use('cli-color');

FactoryAdapter = use('@Konsserto/Component/EntityManager/Adapter/FactoryAdapter');

EntityManager = (function() {
  function EntityManager() {
    this.repositories = [];
    this.objectToPersist = [];
    this.loadAdapter();
  }

  EntityManager.prototype.getRepository = function(repository) {
    var bundle, realRepository, returnRepository, splittedRepository, vendor;
    splittedRepository = repository.split(':');
    vendor = splittedRepository[0];
    bundle = splittedRepository[1];
    realRepository = splittedRepository[2];
    returnRepository = new (use('/src/' + vendor + '/' + bundle + '/Repositories/' + realRepository));
    returnRepository.setAdapter(this.adapter);
    if (this.adapter) {
      returnRepository.setAdapter(this.adapter);
    }
    return returnRepository;
  };

  EntityManager.prototype.loadAdapter = function() {
    this.adapter = FactoryAdapter.getInstance();
    if (!this.adapter) {
      return console.log(cc.red('[WARNING] The driver ' + CONFIG.database.driver + ' is not known by the system, connection to the database no possible'));
    }
  };

  EntityManager.prototype.persist = function(object) {
    object.state = 'managed';
    return this.objectToPersist.push(object);
  };

  EntityManager.prototype.flush = function() {
    var object, _i, _len, _ref, _results;
    _ref = this.objectToPersist;
    _results = [];
    for (_i = 0, _len = _ref.length; _i < _len; _i++) {
      object = _ref[_i];
      if (object.state === 'managed') {
        _results.push(this.getRepositoryFromObject(object).add(object));
      } else {

      }
    }
    return _results;
  };

  EntityManager.prototype.getRepositoryFromObject = function(object) {
    return {};
  };

  return EntityManager;

})();

module.exports = EntityManager;
