// konsserto/lib/src/Konsserto/Component/EntityManager/EntityManager.js
var EntityManager, FactoryAdapter, cc;

cc = use('cli-color');

FactoryAdapter = use('@Konsserto/Component/EntityManager/Adapter/FactoryAdapter');

EntityManager = (function() {
  function EntityManager() {
    this.objectToPersist = [];
    this.loadAdapter();
  }

  EntityManager.prototype.getRepository = function(repository) {
    var bundle, realRepository, returnRepository, splittedRepository, vendor;
    splittedRepository = repository.split(':');
    vendor = splittedRepository[0];
    bundle = splittedRepository[1];
    realRepository = splittedRepository[2];
    returnRepository = new (use('/src/' + vendor + '/' + bundle + '/Repositories/' + realRepository));
    returnRepository.setAdapter(this.adapter);
    if (this.adapter) {
      returnRepository.setAdapter(this.adapter);
    }
    return returnRepository;
  };

  EntityManager.prototype.loadAdapter = function() {
    return this.adapter = FactoryAdapter.getInstance();
  };

  EntityManager.prototype.persist = function(object) {
    object.state = 'managed';
    return this.objectToPersist.push(object);
  };

  EntityManager.prototype.flush = function() {
    var object, repository, _i, _len, _ref, _results;
    _ref = this.objectToPersist;
    _results = [];
    for (_i = 0, _len = _ref.length; _i < _len; _i++) {
      object = _ref[_i];
      if (object.state === 'managed') {
        repository = this.getRepository(object.getRepository());
        if (object.attributes._id.value !== null) {
          _results.push(repository.modifyById(object.attributes._id.value, object));
        } else {
          _results.push(repository.add(object).then(function(newValue) {}));
        }
      } else {

      }
    }
    return _results;
  };

  return EntityManager;

})();

module.exports = EntityManager;
