// konsserto/lib/src/Konsserto/Component/EntityManager/EntityManager.js
var CONFIG, EntityManager, cc;

CONFIG = use('/app/config/config');

cc = use('cli-color');

EntityManager = (function() {
  function EntityManager() {
    this.repositories = [];
    this.objectToPersist = [];
    this.loadAdapter();
  }

  EntityManager.prototype.getRepository = function(repository) {
    var bundle, realRepository, returnRepository, splittedRepository, vendor;
    splittedRepository = repository.split(':');
    vendor = splittedRepository[0];
    bundle = splittedRepository[1];
    realRepository = splittedRepository[2];
    returnRepository = new (use('/src/' + vendor + '/' + bundle + '/Repositories/' + realRepository));
    returnRepository.setAdapter(this.adapter);
    if (this.adapter) {
      returnRepository.setAdapter(this.adapter);
    }
    return returnRepository;
  };

  EntityManager.prototype.loadAdapter = function() {
    if (CONFIG.database) {
      if (!CONFIG.database.host) {
        console.log(cc.red('[ERROR DATABASE] Missing Host'));
        return;
      }
      if (!CONFIG.database.port) {
        console.log(cc.red('[ERROR DATABASE] Missing Host'));
        return;
      }
      if (!CONFIG.database.name) {
        console.log(cc.red('[ERROR DATABASE] Missing database name'));
        return;
      }
      if (!CONFIG.database.user) {
        console.log(cc.red('[ERROR DATABASE] Missing database user'));
        return;
      }
      if (CONFIG.database.driver === 'mysql') {
        return this.adapter = use('@Konsserto/Component/EntityManager/Adapter/MysqlAdapter').getInstance();
      } else {
        if (CONFIG.database.driver === 'mongodb') {
          return this.adapter = use('@Konsserto/Component/EntityManager/Adapter/MongodbAdapter').getInstance();
        } else {
          return console.log(cc.red('[WARNING] The driver ' + CONFIG.database.driver + ' is not known by the system, connection to the database no possible'));
        }
      }
    }
  };

  EntityManager.prototype.persist = function(object) {
    object.state = 'managed';
    return this.objectToPersist.push(object);
  };

  EntityManager.prototype.flush = function() {
    var object, _i, _len, _ref, _results;
    _ref = this.objectToPersist;
    _results = [];
    for (_i = 0, _len = _ref.length; _i < _len; _i++) {
      object = _ref[_i];
      if (object.state === 'managed') {
        _results.push(this.getRepositoryFromObject(object).add(object));
      } else {

      }
    }
    return _results;
  };

  EntityManager.prototype.getRepositoryFromObject = function(object) {
    return {};
  };

  return EntityManager;

})();

module.exports = EntityManager;
