// konsserto/lib/src/Konsserto/Component/EntityManager/Repository.js
var Repository;

Repository = (function() {
  function Repository(collectionName) {
    this.collectionName = collectionName;
  }

  Repository.prototype.getName = function() {
    return this.collectionName;
  };

  Repository.prototype.setAdapter = function(adapter) {
    this.adapter = adapter;
    return this;
  };

  Repository.prototype.getAll = function() {
    return this.adapter.getAll(this.collectionName);
  };

  Repository.prototype.getById = function(id) {
    return this.adapter.getById(this.collectionName, id);
  };

  Repository.prototype.getByCriteria = function(criteria) {
    return this.adapter.getByCriteria(this.collectionName, criteria);
  };

  Repository.prototype.add = function(object) {
    return this.adapter.add(this.collectionName, object);
  };

  Repository.prototype.remove = function(object) {
    return this.adapter.remove(object);
  };

  Repository.prototype.removeById = function(id) {
    return this.adapter.remove(id);
  };

  Repository.prototype.modify = function(objectSource, newObject) {
    return this.adapter.modify(objectSource, newObject);
  };

  Repository.prototype.modifyById = function(id, newObject) {
    return this.adapter.modifyById(this.collectionName, id, newObject);
  };

  return Repository;

})();

module.exports = Repository;
