// konsserto/lib/src/Konsserto/Component/EntityManager/Repository.js
var Repository;

Repository = (function() {
  function Repository() {}

  Repository.prototype.setAdapter = function(adapter) {
    this.adapter = adapter;
  };

  Repository.prototype.getAll = function() {
    return this.adapter.getAll();
  };

  Repository.prototype.getById = function(id, joins) {
    return this.adapter.getById(id, joins);
  };

  Repository.prototype.getByCriteria = function(criteria, joins) {
    return this.adapter.getByCriteria(criteria, joins);
  };

  Repository.prototype.add = function(object) {
    return this.adapter.add(object);
  };

  Repository.prototype.remove = function(object) {
    return this.adapter.remove(object);
  };

  Repository.prototype.removeById = function(id) {
    return this.adapter.remove(id);
  };

  Repository.prototype.modify = function(objectSource, newObject) {
    return this.adapter.modify(objectSource, newObject);
  };

  Repository.prototype.modifyById = function(id, newObject) {
    return this.adapter.modify(id, newObject);
  };

  return Repository;

})();

module.exports = Repository;
